src/InvariantTheory/invariant_rings.jl

```diff
@@ -1,130 +1,166 @@
-      RS = Oscar.singular_ring(R)
-      singular_matrices = Vector{Singular.smatrix{elem_type(RS)}}(undef, length(IR.matrices))
-      for i = 1:length(IR.matrices)
-         singular_matrices[i] = map_entries(RS, IR.matrices[i])
-      end
-
+      R = polynomial_ring(IR)
+      singular_matrices = _action_singular(IR)
       rey, mol = Singular.LibFinvar.reynolds_molien(singular_matrices...)
       IR.reynolds_singular = rey
       IR.molien_singular = mol
    end
    return IR.reynolds_singular, IR.molien_singular
 end
 
-function primary_invariants_via_singular(IR::InvariantRing)
-   @assert characteristic(IR.field) == 0
+function primary_invariants_via_singular(IR::InvRing)
    if !isdefined(IR, :primary_singular)
-      rey, mol = reynolds_molien_via_singular(IR)
-      P = Singular.LibFinvar.primary_char0(rey, mol)
-      p = Vector{elem_type(IR.poly_ring)}()
-      for i = 1:ncols(P)
-         push!(p, IR.poly_ring(P[1, i]))
+      if ismodular(IR)
+         rey, mol = reynolds_molien_via_singular(IR)
+         P = Singular.LibFinvar.primary_char0(rey, mol)
+         R = polynomial_ring(IR)
+         p = Vector{elem_type(R)}()
+         for i = 1:ncols(P)
+            push!(p, R(P[1, i]))
+         end
+         IR.primary_singular = P
+         IR.primary = p
+      else
+         P = Singular.LibFinvar.primary_invariants(_action_singular(IR)...)[1]
+         R = polynomial_ring(IR)
+         p = Vector{elem_type(R)}()
+         for i = 1:ncols(P)
+           push!(p, R(P[1, i]))
+         end
+         IR.primary_singular = P
+         IR.primary = p
       end
-      IR.primary_singular = P
-      IR.primary = p
    end
    return IR.primary
 end
 
-function primary_invariants(IR::InvariantRing)
+function primary_invariants(IR::InvRing)
    if !isdefined(IR, :primary)
       return primary_invariants_via_singular(IR)
    end
-   return IR.primary
+   return copy(IR.primary)
 end
 
-function secondary_invariants_via_singular(IR::InvariantRing)
+function secondary_invariants_via_singular(IR::InvRing)
    if !isdefined(IR, :secondary)
       rey, mol = reynolds_molien_via_singular(IR)
       primary_invariants_via_singular(IR)
       P = IR.primary_singular
       S, IS = Singular.LibFinvar.secondary_char0(P, rey, mol)
-      s = Vector{elem_type(IR.poly_ring)}()
+      R = polynomial_ring(IR)
+      s = Vector{elem_type(R)}()
       for i = 1:ncols(S)
-         push!(s, IR.poly_ring(S[1, i]))
+         push!(s, R(S[1, i]))
       end
-      is = Vector{elem_type(IR.poly_ring)}()
+      is = Vector{elem_type(R)}()
       for i = 1:ncols(IS)
-         push!(is, IR.poly_ring(IS[1, i]))
+         push!(is, R(IS[1, i]))
       end
       IR.secondary = s
       IR.irreducible_secondary = is
    end
    return IR.secondary
 end
 
-function secondary_invariants(IR::InvariantRing)
+function secondary_invariants(IR::InvRing)
    if !isdefined(IR, :secondary)
       return secondary_invariants_via_singular(IR)
    end
    return IR.secondary
```

Comment:
Do we also return a copy here?

Reply:
Yeah, right.

Reply:
I changed the logic a bit so to make a copy only in the user functions.